Add more logging for certificate exchange/validation

caesar1995 opened the pull request:
1. Added Logging when server request a client certificate.
2. Added logging when client is unable to find matching certificate.
3. Made a log entry in `VerifyRemoteCertificate()` more clear.
4. Corrected a typo.

Verified that we already logged why the certificate is rejected (in `LogCertificateValidation()`). Wrote some tests and used PerfView to examine the tracing manually.

davidsh commented:
You should consider adding logging to this class:

https://github.com/dotnet/corefx/blob/master/src/Common/src/System/Net/Security/CertificateHelper.cs

All the places where we reject/skip the certificate is a good place to log.

caesar1995 commented:
Good to know about CertificateHelper class. I was looking into `System.Net.Security` and forget to check tracing for HttpClientHandler.

I have addressed the feedback and did manually verification on the new logging.

rmkerr commented on src/Common/src/System/Net/Security/CertificateHelper.cs:
```diff
@@ -38,15 +38,27 @@ internal static X509Certificate2 GetEligibleClientCertificate(X509Certificate2Co
                 {
+                    if (NetEventSource.IsEnabled)
+                    {
+                        NetEventSource.Info(candidateCerts, $"Skipping current X509Certificate2 {cert.GetHashCode()} since it doesn't has private key.");
```
nit: doesn't have a private key

caesar1995 replied:
Will change.

davidsh commented:
Based on the logging being added such as:

```c#
NetEventSource.Info(candidateCerts, $"Skipping current X509Certificate2 {cert.GetHashCode()} since it doesn't have private key.");
```

how does a developer determine which certificate of theirs (from a possible set of multiple certificates) matches this: `cert.GetHashCode()`?  Perhaps in addition to logging `GetHashCode`, the 'Subject' and/or 'Thumbprint' of the certificate should be logged as well.  That would make it easier to figure out which cert it is.  The GetHashCode() is only useful for tracking the object's .NET identity thru all the logging. But at some point, that value isn't useful when looking in the certificate store, etc.  Something like 'Subject' and/or 'Thumbprint' is useful since you can search for in the store.

davidsh commented on src/System.Net.Security/src/System/Net/Security/SecureChannel.cs:
```diff
@@ -1005,7 +1009,7 @@ internal bool VerifyRemoteCertificate(RemoteCertValidationCallback remoteCertVal
                 {
                     if (NetEventSource.IsEnabled)
-                        NetEventSource.Exit(this, "(no remote cert)", !_sslAuthenticationOptions.RemoteCertRequired);
+                        NetEventSource.Exit(this, $"No remote certificate received. RemoteCertRequired: {RemoteCertRequired}");
```
why is this changing from `!_sslAuthenticationOptions.RemoteCertRequired` to `RemoteCertRequired`?  Is it an equivalent value?  It looks like it might be missing the "!" operator ?

caesar1995 replied:
I changed it because I found the log message is confusing, and I want to improve it.

The `VerifyRemoteCertificate()` function tries to get the remote certificate (the certificate from other side). When there is no remote certificate received, it makes more sense to log whether current host (client/server) requires remote certificate or not.

```c#
        internal bool RemoteCertRequired
        {
            get
            {
                return _sslAuthenticationOptions.RemoteCertRequired;
            }
        }
```

I apply both changes just to compare the output message side-by-side.
<img width="724" alt="capture" src="https://user-images.githubusercontent.com/8537784/43018233-d4b5ad76-8c0d-11e8-940d-3750f9eb35d0.PNG">

davidsh replied:
>When there is no remote certificate received, it makes more sense to log whether current host (client/server) requires remote certificate or not.

I don't agree with this completely.  There is a difference in whether the "local" side is requesting a certificate or not from the "remote" side vs. whether the "remote" side provided one at all.  In fact, it probably makes sense to log both pieces of information.  Most connection failures are because the "local" side requested a cert and the "remote" side did not send one.

caesar1995 replied:
I think current modified logging include both pieces of information.

> whether the "local" side is requesting a certificate or not from the "remote" side

This information is in `RemoteCertRequired`.

> whether the "remote" side provided one at all

This information is in `No remote certificate received`.

Did I misunderstand something?
